test/CMakeLists.txt

```diff
@@ -13,6 +26,6 @@ add_executable(menoh_test
     #vgg16.cpp
 
 )
-target_link_libraries(menoh_test gtest_main menoh_test_target)
+target_link_libraries(menoh_test gtest_main menoh_test_target ${PROTOBUF_LIBRARY})
```

Comment:
This causes following link error on MINGW when `LINK_STATIC_LIBPROTOBUF=ON`.
This is because `libmenoh_test_target` in that case contains libprotobuf and exports their symbols causing duplicated symbols.

```
[ 79%] Linking CXX executable menoh_test.exe
C:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/8.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: ../protobuf-3.6.1/lib/libprotobuf.a(coded_stream.o):coded_stream.cc:(.text+0x58): multiple definition of `google::protobuf::io::CodedInputStream::~CodedInputStream()'; ../menoh/libmenoh_test_target.dll.a(d003033.o):(.text+0x0): first defined here
C:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/8.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: ../protobuf-3.6.1/lib/libprotobuf.a(coded_stream.o):coded_stream.cc:(.text+0x414): multiple definition of `google::protobuf::io::CodedInputStream::SetTotalBytesLimit(int)'; ../menoh/libmenoh_test_target.dll.a(d003001.o):(.text+0x0): first defined here
C:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/8.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: ../protobuf-3.6.1/lib/libprotobuf.a(coded_stream.o):coded_stream.cc:(.text+0x1d3e): multiple definition of `google::protobuf::io::CodedInputStream::Refresh()'; ../menoh/libmenoh_test_target.dll.a(d003028.o):(.text+0x0): first defined here
C:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/8.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: ../protobuf-3.6.1/lib/libprotobuf.a(zero_copy_stream_impl.o):zero_copy_stream_impl.cc:(.text+0xaf4): multiple definition of `google::protobuf::io::IstreamInputStream::IstreamInputStream(std::istream*, int)'; ../menoh/libmenoh_test_target.dll.a(d003109.o):(.text+0x0): first defined here
C:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/8.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: ../protobuf-3.6.1/lib/libprotobuf.a(zero_copy_stream_impl.o):zero_copy_stream_impl.cc:(.text+0xc74): multiple definition of `google::protobuf::io::IstreamInputStream::CopyingIstreamInputStream::~CopyingIstreamInputStream()'; ../menoh/libmenoh_test_target.dll.a(d003104.o):(.text+0x0): first defined here
C:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/8.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: ../protobuf-3.6.1/lib/libprotobuf.a(zero_copy_stream_impl_lite.o):zero_copy_stream_impl_lite.cc:(.text+0xde6): multiple definition of `google::protobuf::io::CopyingInputStreamAdaptor::~CopyingInputStreamAdaptor()'; ../menoh/libmenoh_test_target.dll.a(d003164.o):(.text+0x0): first defined here
collect2.exe: error: ld returned 1 exit status
make[2]: *** [test/CMakeFiles/menoh_test.dir/build.make:210: test/menoh_test.exe] Error 1
make[1]: *** [CMakeFiles/Makefile2:164: test/CMakeFiles/menoh_test.dir/all] Error 2
make: *** [Makefile:130: all] Error 2
Command exited with code 2
```

Simple ad-hoc fix  would be adding `${PROTOBUF_LIBRARY}` only when `LINK_STATIC_LIBPROTOBUF=OFF`, but I think we need to consider linking interface of `libmenoh_test_target` more seriously (applying version script, link interface keyword in CMake, etc.), @okapies.
OT: I'm also worrying about the usage of `-static-libgcc` and `-static-libstdc++` without version script.

Reply:
The tests requires to access some symbols but not others. I agree with you to add a version script for `libmenoh_test_target`.

Reply:
Hmm,

https://github.com/pfnet-research/menoh/blob/a79c0593036cb36c311d1971159c609d73680264/test/operator.cpp#L38-L42

Implementation of `onnx::TensorProto` in `libmenoh_test_target` is used with `gpio::CodedInputStream`.

When `LINK_STATIC_LIBPROTOBUF=ON` and `libprotobuf` copy is inside `libmenoh_test_target`, `onnx::TensorProto` is compiled to work with that `libprotobuf` copy and using a `gpio::CodedInputStream` coming from another `libprotobuf` copy looks dangerous.
Is this correct?

Reply:
As you said, sharing protobuf objects between two `libprotobuf` runtimes potentially breaks our code because these objects tend to have references to the runtime. cf. https://github.com/okapies/protobuf-shareddtor-crash

Otherwise, I think it is not problem if the created object is created in the test and run by the runtime for test binary.

Reply:
Let's fix the issue simply as https://github.com/msakai/menoh/commit/64afa29312c25adc85d554d125ffa7a9e08702ab for now, and do more proper fix later in different branch.

@okdshin, if it's okay, could you add the commit to this branch.

Reply:
It sounds good. This PR is very important for later PRs. I'm agree with merging it anyway.　Thank you @msakai and @okapies